### PR TITLE
HOSTEDCP-1518: Preserve container resource requests and limits

### PR DESCRIPTION
**Pull request by lucaspalm**
**What this PR does / why we need it**:
There are a few components leftover from the changes in https://github.com/openshift/hypershift/pull/3120 that still do not exhibit the expected resource preservation behavior.  The changes in this PR address these issues to ensure all control plane components preserve any externally sourced modifications to their resource requests/limits during their respective reconciliations.

**Which issue(s) this PR fixes**:
https://issues.redhat.com/browse/HOSTEDCP-1518

### CODE REVIEW

**Comment by rtheis** [comment]
@lucaspalm verification is failing, ptal

**Review by rtheis — APPROVED** [review]
/lgtm

**Review by stevekuznetsov — COMMENTED** [review]
There are a couple good examples of using golden fixture files to test these `Reconcile*()` functions - could you please add such a test to these modules and show how the functions now correctly honor previously set resource requests and/or limits?

**Comment by lucaspalm** [comment]
> There are a couple good examples of using golden fixture files to test these `Reconcile*()` functions - could you please add such a test to these modules and show how the functions now correctly honor previously set resource requests and/or limits?

@stevekuznetsov Can you point me to one such example?

**Comment by stevekuznetsov** [comment]
Sure, [here](https://github.com/openshift/hypershift/blob/9416177c4095b8119017fca3904e3d3d547c88c5/control-plane-operator/controllers/hostedcontrolplane/pkioperator/reconcile_test.go#L18) is one example/

**Comment by lucaspalm** [comment]
@stevekuznetsov I've added tests in https://github.com/openshift/hypershift/pull/3864/commits/3e2b6b9fc19516a8706f5449f0ed7537a24c4ca9 that seem to work.  Let me know if thats what you had in mind.

Thanks.

File: control-plane-operator/controllers/hostedcontrolplane/cno/clusternetworkoperator_test.go [hunk 1]
```diff
@@ -35,7 +36,29 @@ func TestReconcileDeployment(t *testing.T) {
 				tc.params.ReleaseVersion = "4.11.0"
 			}
 
-			dep := &appsv1.Deployment{}
```
Comment (stevekuznetsov) [inline control-plane-operator/controllers/hostedcontrolplane/cno/clusternetworkoperator_test.go:38]: The reason I suggested using the golden file approach is that it's much more ergonomic - I do not recommend it, but if you want to keep this format, though, please make the deployment a parameter for the test case and only test this functionality in one test case. You'll want to test the output of the reconcile method without pre-existing values as well.

File: control-plane-operator/controllers/hostedcontrolplane/cno/clusternetworkoperator_test.go [hunk 2]
```diff
@@ -53,6 +76,14 @@ func TestReconcileDeployment(t *testing.T) {
 					strconv.FormatBool(tc.expectProxyAPIServerAddress),
 					strconv.FormatBool(hasProxyAPIServerAddress))
 			}
+
+			// Verify the existing resources were preserved
+			if dep.Spec.Template.Spec.Containers[0].Resources.Requests.Cpu().MilliValue() != 100 ||
```
Comment (stevekuznetsov) [inline control-plane-operator/controllers/hostedcontrolplane/cno/clusternetworkoperator_test.go:81]: Try breaking your test and see what the error output looks like. Is it useful to developers and quick to determine? Try using the golden file approach and/or `cmp.Diff()` and compare those outputs - the latter two options provide clear, minimal diffs that make it easy to a) provide test input fixures, b) declare expected output, and c) clarify what errored, how, and where.

Comment (stevekuznetsov) [reply control-plane-operator/controllers/hostedcontrolplane/cno/clusternetworkoperator_test.go:81]: More succinctly - if I were to break this, I would not want to know that "some or all..." - a good test explicitly tells me which I broke, and makes it quick to determine where that failure is and how I broke it.

File: control-plane-operator/controllers/hostedcontrolplane/cno/clusternetworkoperator_test.go (hunk 1 above)
Comment (lucaspalm) [reply control-plane-operator/controllers/hostedcontrolplane/cno/clusternetworkoperator_test.go:38]: I think I understand what you are saying now about the golden file testing.  See my latest update in https://github.com/openshift/hypershift/pull/3864/commits/6e62f5b3b65cc40a22f43776a41f0cab46ff0b75.

**Review by stevekuznetsov — APPROVED** [review]

**Review by rtheis — APPROVED** [review]
/lgtm

**Comment by rtheis** [comment]
@lucaspalm looks like a manual cherry-pick is required for this.

**Comment by lucaspalm** [comment]
Manual cherry-pick created... https://github.com/openshift/hypershift/pull/4032

**Pull request merged**